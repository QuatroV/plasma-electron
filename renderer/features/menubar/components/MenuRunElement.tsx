import { useRef, useState } from "react";
import { ImSpinner9 } from "react-icons/im";
import { IoMdArrowDropdown } from "react-icons/io";

import Dropdown from "../../../components/Dropdown";
import useOnClickOutside from "../../../hooks/useOnClickOutside";
import clsxm from "../../../utils/clsxm";
import useDebug from "../../debug/hooks/useDebug";
import useRunFile from "../hooks/useRunFile";

const MenuRunElement = (props) => {
  const { runFile, currentFileCanBeExecuted, isExecuting } = useRunFile();

  const { startDebug } = useDebug();

  const BuildAndRunCurrentFileOption = () => (
    <div
      onClick={runFile}
      title={!currentFileCanBeExecuted && "Wrong type of file"}
      className={
        !currentFileCanBeExecuted && "cursor-not-allowed text-gray-500"
      }
    >
      Build And Run Current File{" "}
    </div>
  );

  const BuildAndRunFileOption = () => <div>Build And Run File...</div>;

  const BuildAndRunProjectOption = () => <div>Build And Run Project</div>;

  const RunCurrentFileOption = () => (
    <div
      onClick={() => null}
      title={!currentFileCanBeExecuted && "Wrong type of file"}
      className={
        !currentFileCanBeExecuted && "cursor-not-allowed text-gray-500"
      }
    >
      Run Current File
    </div>
  );

  const RunFileOption = () => <div>Run File...</div>;

  const RunProjectOption = () => <div>Run Project</div>;

  const RunFileWithDebugOption = () => (
    <div onClick={() => startDebug()}>Run File With Debug...</div>
  );

  const options = [
    BuildAndRunCurrentFileOption,
    BuildAndRunFileOption,
    BuildAndRunProjectOption,
    "divider",
    RunCurrentFileOption,
    RunFileOption,
    RunProjectOption,
    "divider",
    RunFileWithDebugOption,
  ];

  const [dropdownOpen, setDropdownOpen] = useState(false);

  const [shakeEffect, setShakeEffect] = useState(false);

  const dropdownListRef = useRef(null);

  useOnClickOutside(dropdownListRef, () => setDropdownOpen(false));

  const handleClickRunButton = () => {
    if (currentFileCanBeExecuted) {
      runFile();
    } else {
      setShakeEffect(true);
    }
  };

  return (
    <div
      className={clsxm(
        `non-draggable font-rubik flex items-stretch rounded-lg  border bg-white transition-all hover:shadow-lg`,
        shakeEffect ? "animate-fast-shake" : "",
      )}
      onAnimationEnd={() => setShakeEffect(false)}
      {...props}
    >
      <div
        onClick={handleClickRunButton}
        title={!currentFileCanBeExecuted && "Wrong type of file"}
        className={clsxm(
          " flex h-[30px] items-center rounded-l-lg border-r px-2 py-1 ",
          currentFileCanBeExecuted
            ? "cursor-pointer active:bg-gray-300 active:shadow-inner"
            : "cursor-not-allowed",
        )}
      >
        Run
      </div>
      <div
        className={clsxm(
          "relative flex w-full cursor-pointer items-center rounded-r-lg px-1 py-1 active:bg-gray-200 active:shadow-inner",
        )}
      >
        {isExecuting ? (
          <ImSpinner9 color="#727272" className="animate-spin" size={18} />
        ) : (
          <Dropdown
            options={options || []}
            ref={dropdownListRef}
            onClick={() => setDropdownOpen(true)}
            dropdownOpen={dropdownOpen}
            dropdownStyles="left-0"
          >
            <div className="h-full">
              <IoMdArrowDropdown size={18} />
            </div>
          </Dropdown>
        )}
      </div>
    </div>
  );
};

export default MenuRunElement;
