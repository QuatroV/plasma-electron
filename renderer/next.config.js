module.exports = {
  webpack: (config, { isServer }) => {
    if (!isServer) {
      config.target = 'electron-renderer';
    }
    
    config.module.rules.push({
      test: /\.svg$/i,
      issuer: /\.[jt]sx?$/,
      use: ['@svgr/webpack'],
    })

    return config;
  },
};
