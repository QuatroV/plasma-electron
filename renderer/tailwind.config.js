/** @type {import('tailwindcss').Config} */
module.exports = {
  content: [
    './renderer/pages/**/*.{js,ts,jsx,tsx}',
    './renderer/components/**/*.{js,ts,jsx,tsx}',
    './renderer/features/**/*.{js,ts,jsx,tsx}',
  ],
  theme: {
    extend: {
      backgroundImage: {
        'gradient-radial': 'radial-gradient(var(--tw-gradient-stops))',
      },
      keyframes: {
        "slow-appear": {
          '0%': { transform: 'scale(0.1, 1)', transform: 'translate(-5%, 0)', opacity: "0.1" },
          '100%': { transform: 'scale(1, 1)', transform: 'translate(0, 0)', opacity: "1" },
        },
        "slow-appear-no-opacity": {
          '0%': { transform: 'scale(0.1, 1)', transform: 'translate(-5%, 0)'},
          '100%': { transform: 'scale(1, 1)', transform: 'translate(0, 0)'},
        },
        "roll": {
          from: { transform: 'rotate(0deg)'},
          to: { transform: 'rotate(-360deg)'},
        },
        "shake": {
          "10%":{
            transform: "translate3d(-1px, 0, 0)",
          },
          
          "20%":{
            transform: "translate3d(2px, 0, 0)",
          },
        
          "30%":{
            transform: "translate3d(-4px, 0, 0)",
          },
        
          "40%":{
            transform: "translate3d(4px, 0, 0)",
          },

          "50%":{
            transform: "translate3d(-4px, 0, 0)",
          },
        
          "60%":{
            transform: "translate3d(4px, 0, 0)",
          },

          "70%":{
            transform: "translate3d(4px, 0, 0)",
          },

          "80%":{
            transform: "translate3d(2px, 0, 0)",
          },

          "90%":{
            transform: "translate3d(-1px, 0, 0)",
          },
          
        },
        "scale": {
          from: { transform: 'scale(0)'},
          to: { transform: 'scale(1)'},
        }
      },
      animation: {
        'slow-appear': 'slow-appear 0.25s linear',
        "slow-appear-no-opacity": 'slow-appear-no-opacity 0.25s linear',
        'one-roll': 'roll .25s linear',
        'fast-shake': 'shake .25s linear',
        'scale': 'scale .12s linear',
      },
    },
    fontFamily: {
      rubik: ['Rubik', 'sans-serif'],
    }
  },
  plugins:[
    require('@tailwindcss/container-queries'),
  ],
};
