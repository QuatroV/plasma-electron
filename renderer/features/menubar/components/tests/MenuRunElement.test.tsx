import ReactDOM from "react-dom";
import renderer from "react-test-renderer";

import MenuRunElement from "../MenuRunElement";

describe("MenuRunElement", () => {
  // Hack, see link: https://github.com/facebook/react/issues/11565#issuecomment-380143358
  beforeAll(() => {
    //@ts-ignore
    ReactDOM.createPortal = jest.fn((element, node) => {
      return element;
    });
  });

  afterEach(() => {
    //@ts-ignore
    ReactDOM.createPortal.mockClear();
  });
  it("renders correctly", () => {
    const tree = renderer.create(<MenuRunElement />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
