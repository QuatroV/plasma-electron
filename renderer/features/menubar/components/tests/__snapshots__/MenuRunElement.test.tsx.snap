// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MenuRunElement renders correctly 1`] = `
<div
  className="non-draggable font-rubik flex items-stretch rounded-lg border bg-white transition-all hover:shadow-lg"
  onAnimationEnd={[Function]}
>
  <div
    className="flex h-[30px] items-center rounded-l-lg border-r px-2 py-1 cursor-not-allowed"
    onClick={[Function]}
    title="Wrong type of file"
  >
    Run
  </div>
  <div
    className="relative flex w-full cursor-pointer items-center rounded-r-lg px-1 py-1 active:bg-gray-200 active:shadow-inner"
  >
    <div
      className="flex flex-col justify-center"
    >
      <div
        className="h-full"
        onClick={[Function]}
      >
        <div
          className="h-full"
        >
          <svg
            fill="currentColor"
            height={18}
            stroke="currentColor"
            strokeWidth="0"
            style={
              {
                "color": undefined,
              }
            }
            viewBox="0 0 512 512"
            width={18}
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M128 192l128 128 128-128z"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>
</div>
`;
